src/db/models.rs

```diff
@@ -173,31 +179,59 @@ impl DBManager {
     }
 
     pub fn get_bso(&self, collection_id: i64, bso_id: &str) -> Result<Option<BSO>, DieselError> {
-        let result = bso::table
+        bso::table
             .select(bso::table::all_columns())
             .filter(bso::collection_id.eq(&collection_id))
             .filter(bso::id.eq(&bso_id))
             .filter(bso::expiry.ge(&ms_since_epoch()))
-            .first::<BSO>(&self.conn);
-        match result {
-            Ok(bso) => Ok(Some(bso)),
-            Err(NotFound) => Ok(None),
-            Err(e) => Err(e),
-        }
+            .first::<BSO>(&self.conn)
+            .optional()
+    }
+
+    pub fn get_bso_modified(&self, collection_id: i64, bso_id: &str) -> Result<i64, DieselError> {
+        let last_modified = bso::table
+            .select(bso::last_modified)
+            .filter(bso::collection_id.eq(&collection_id))
+            .filter(bso::id.eq(&bso_id))
+            .filter(bso::expiry.ge(&ms_since_epoch()))
+            .first(&self.conn)
+            .optional()?;
+        // XXX: returning 0 when not found or prefer Option?
+        Ok(last_modified.unwrap_or(0))
+    }
+
+    pub fn delete_bso(&self, collection_id: i64, bso_id: &str) -> Result<i64, DieselError> {
+        self.delete_bsos(collection_id, &[bso_id])
+    }
+
+    pub fn delete_bsos(&self, collection_id: i64, bso_id: &[&str]) -> Result<i64, DieselError> {
+        self.conn.transaction(|| {
+            delete(bso::table)
+                .filter(bso::collection_id.eq(&collection_id))
+                .filter(bso::id.eq_any(bso_id))
+                .execute(&self.conn)?;
+
+            let last_modified = ms_since_epoch();
```

Comment:
There's an interesting architectural choice here, which maybe won't ever come up in practice but that I wanted to mention.

From the point of view of the Firefox Sync client, an individual request to the syncstorage API is supposed to be atomic and occur at a specific instant in time.  For example, if the client does a `POST /storage/collection` and sends three items, then it expects that:

* The last-modified time of each of those three items will be the same
* The new last-modified time of the collection will be the same as those items just added
* The new last-modified time of the storage as a whole will be the same as that of the collection
* The response will include an `X-Last-Modified` header that gives this new timestamp

So is the "instant" at which all those changes occur a property of the individual call into the backend storage API (as it is here)?  Or should it be a property of the request as a whole?  Should some higher-level code be passing in a notion of "this is the current timestamp" in order to help ensure this consistency?

If each web API request corresponds to exactly one backend storage API request, then it probably doesn't matter.  In the python code, I believe we fix a timestamp when we open a database transaction, and use it for all operations done inside that transaction.  I don't know what the best approach will be for this new code.

Just something to think about as this code grows :-)

Reply:
I did notice this difference in the Python version vs go (go timestamps in the db layer). We'll keep this in mind, the latter might be simpler but we should only do it if it works in the same consistent way as the Python version.